### PR TITLE
Add auto locate key support

### PR DESCRIPTION
**Pull request by Myzel394**
Hey @vsajip thanks for your awesome package! I'd like to add auto locate key support (retrieving keys from WKD for example).

I managed to add support for retrieving the fingerprint, but I'd like to include all the information `gpg` provides.
Here's an example of what `gpg` outputs to the console when autolocating a key:

```shell
$ gpg --auto-key-locate wkd --locate-key no-reply@protonmail.com
pub   rsa2048 2021-07-14 [SC]
      90E619A84E85330A692F6D81A655882018DBFA9D
uid           [ unknown] no-reply@protonmail.com <no-reply@protonmail.com>
sub   rsa2048 2021-07-14 [E]
```

You can see that it tells you when the key was generated and what type it is. I'll need some help from you regarding this information. How can I extract it? I'm using `StatusHandler` to extract the fingerprint and the email, but how do I find out the other stuff?

### CODE REVIEW

**Comment by Myzel394** [comment]
will solve #117

**Comment by vsajip** [comment]
> How can I extract it?

The same way that e.g. `list_keys` does. Note that any patch will need to check the `gpg` version and behave appropriately (see my above comment on the issue about key location) and work/be tested across platforms.

**Comment by Myzel394** [comment]
Hey @vsajip by default you add `--with-colons` to the args. I noticed that this option changes the output drastically (stripping away lots of information), is there any particular reason why we need this?

**Comment by vsajip** [comment]
> is there any particular reason why we need this? [`--with-colons`]

It's there to facilitate machine parsing of the information - as everything is delimited by colons, there's no need to guess how things are laid out.

**Comment by vsajip** [comment]
If information is lost when `--with-colons` is used for your use case, you might need to take it up with the GnuPG project.

**Comment by Myzel394** [comment]
@vsajip after investigating the output a bit more, it seems as if the data is encoded into a different format (timestamp instead of human readable dates for example). I searched through the gnupg docs but I can't find any information on how the other data is encoded. What resources did you use to parse the output of the other functions from your library?

**Comment by vsajip** [comment]
> What resources did you use to parse the output of the other functions from your library?

The `DETAILS` file from the GnuPG sources.

**Comment by Myzel394** [comment]
@vsajip thank you, those docs helped very much. I've now added the parsers. However, I'm not a pgp keys expert myself, but it seems as if there are multiple different fingerprints returned for the same key - i decided to simply use the first key, as it seems to be the correct one. Can you take a look over the code?

**Comment by Myzel394** [comment]
@vsajip Are there any news on this?

**Comment by vsajip** [comment]
Sorry, been busy on other things. Will get to it as soon as I can.

File: gnupg.py [hunk 1]
```diff
@@ -1000,6 +1002,55 @@ def handle_status(self, key, value):
...
 
+class AutoLocateKey(StatusHandler):
+    """
+    This class handles status messages during key auto-locating.
+    """
+    fingerprint: str
+    key_length: int
+    created_at: date
+    email: str
+    email_real_name: str
```
Comment (vsajip) [inline gnupg.py]:
```suggestion
    fingerprint: str
    key_length: int
    created_at: date
    email: str
    email_real_name: str
    """
```

**Comment by vsajip** [comment]
Still getting errors when testing locally. From my log:
```console
2024-01-23 10:33:57,830 DEBUG gnupg      Thread-3   1235 gpg: error retrieving 'no-reply@protonmail.com' via WKD: Invalid URI
2024-01-23 10:33:58,026 DEBUG gnupg      Thread-3   1235 gpg: error retrieving 'no-reply@protonmail.com' via NTDS: No keyserver available
2024-01-23 10:33:58,026 DEBUG gnupg      Thread-3   1235 gpg: error retrieving 'no-reply@protonmail.com' via LDAP: Not implemented
2024-01-23 10:33:58,026 DEBUG gnupg      Thread-3   1235 gpg: error retrieving 'no-reply@protonmail.com' via DNS CERT: No name
2024-01-23 10:33:58,031 DEBUG gnupg      Thread-3   1235 gpg: error retrieving 'no-reply@protonmail.com' via DANE: No name
```

File: gnupg.py [hunk 2]
```diff
@@ -1869,6 +1921,30 @@ def search_keys(self, query, keyserver='pgp.mit.edu', extra_args=None):
                 getattr(result, keyword)(L)
         return result
 
+    def auto_locate_key(self, email, mechanisms=None, **kwargs) -> AutoLocateKey:
```
Comment (vsajip) [inline gnupg.py]:
```suggestion
    def auto_locate_key(self, email, mechanisms=None, **kwargs):
```

**Comment by vsajip** [comment]
Thanks very much for this patch. I've incorporated the changes but made some local edits to do with testing locally, so I won't merge this PR - but your changes are incorporated in 3df9074.

**Comment by Myzel394** [comment]
No problem, I'm glad this feature got added!

**Pull request closed without merging**